File: apps/android/app/src/main/java/org/llamenos/hotline/ui/conversations/ConversationsViewModel.kt
package org.llamenos.hotline.ui.conversations

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.llamenos.hotline.api.ApiService
import org.llamenos.hotline.api.SessionState
import org.llamenos.hotline.api.WebSocketService
import org.llamenos.hotline.crypto.CryptoService
import org.llamenos.hotline.model.Conversation
import org.llamenos.hotline.model.ConversationMessage
import org.llamenos.hotline.model.ConversationsListResponse
import org.llamenos.hotline.model.CreateMessageEnvelope
import org.llamenos.hotline.model.DecryptedMessage
import org.llamenos.hotline.model.LlamenosEvent
import org.llamenos.hotline.model.MessagesListResponse
import org.llamenos.hotline.model.SendMessageRequest
import org.llamenos.hotline.model.User
import org.llamenos.hotline.model.UsersListResponse
import javax.inject.Inject

/**
 * Filter options for the conversations list.
 */
enum class ConversationFilter(val queryParam: String) {
    ACTIVE("active"),
    CLOSED("closed"),
    ALL("all"),
}

data class ConversationsUiState(
    val conversations: List<Conversation> = emptyList(),
    val isLoading: Boolean = false,
    val isRefreshing: Boolean = false,
    val error: String? = null,
    val filter: ConversationFilter = ConversationFilter.ACTIVE,
    val totalConversations: Int = 0,
    val totalUnread: Int = 0,

    // Detail view
    val selectedConversation: Conversation? = null,
    val messages: List<DecryptedMessage> = emptyList(),
    val isLoadingMessages: Boolean = false,
    val messagesError: String? = null,
    val totalMessages: Int = 0,

    // Sending
    val isSending: Boolean = false,
    val sendError: String? = null,

    // Search
    val searchQuery: String = "",

    // Actions
    val showAssignDialog: Boolean = false,
    val assignableVolunteers: List<User> = emptyList(),
    val isLoadingVolunteers: Boolean = false,
)

/**
 * ViewModel for the Conversations feature.
 *
 * Manages fetching, decrypting, and sending E2EE messages within conversations.
 * Each message is encrypted with per-message forward secrecy: a unique random
 * symmetric key per message, ECIES-wrapped for the assigned volunteer and admins.
 *
 * Subscribes to WebSocket events for real-time message delivery and conversation
 * status updates.
 */
@HiltViewModel
class ConversationsViewModel @Inject constructor(
    private val apiService: ApiService,
    private val cryptoService: CryptoService,
    private val webSocketService: WebSocketService,
    private val sessionState: SessionState,
) : ViewModel() {

    private val _uiState = MutableStateFlow(ConversationsUiState())
    val uiState: StateFlow<ConversationsUiState> = _uiState.asStateFlow()

    init {
        loadConversations()
        subscribeToEvents()
    }

    /**
     * Subscribe to real-time conversation and message events from the WebSocket.
     */
    private fun subscribeToEvents() {
        viewModelScope.launch {
            webSocketService.typedEvents.collect { event ->
                when (event) {
                    is LlamenosEvent.MessageNew -> {
                        // If we are viewing this conversation, reload messages
                        val selected = _uiState.value.selectedConversation
                        if (selected != null && selected.id == event.conversationId) {
                            loadMessages(event.conversationId)
                        }
                        // Refresh conversation list for unread count update
                        loadConversations()
                    }
                    is LlamenosEvent.ConversationAssigned,
                    is LlamenosEvent.ConversationClosed -> {
                        loadConversations()
                    }
                    else -> { /* ignore non-conversation events */ }
                }
            }
        }
    }

    /**
     * Load conversations from the API, filtered by the current status filter.
     */
    fun loadConversations() {
        viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isLoading = it.conversations.isEmpty(),
                    isRefreshing = it.conversations.isNotEmpty(),
                    error = null,
                )
            }

            try {
                val filter = _uiState.value.filter
                val response = apiService.request<ConversationsListResponse>(
                    "GET",
                    "/api/conversations?status=${filter.queryParam}",
                )

                val totalUnread = response.conversations.sumOf { it.unreadCount }

                _uiState.update {
                    it.copy(
                        conversations = response.conversations,
                        isLoading = false,
                        isRefreshing = false,
                        totalConversations = response.total,
                        totalUnread = totalUnread,
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        isRefreshing = false,
                        error = e.message ?: "Failed to load conversations",
                    )
                }
            }
        }
    }

    /**
     * Change the status filter and reload.
     */
    fun setFilter(filter: ConversationFilter) {
        _uiState.update { it.copy(filter = filter) }
        loadConversations()
    }

    /**
     * Pull-to-refresh.
     */
    fun refresh() {
        loadConversations()
    }

    /**
     * Select a conversation and load its messages.
     */
    fun openConversation(conversation: Conversation) {
        _uiState.update {
            it.copy(
                selectedConversation = conversation,
                messages = emptyList(),
                isLoadingMessages = true,
                messagesError = null,
            )
        }
        loadMessages(conversation.id)
        markAsRead(conversation.id)
    }

    /**
     * Clear the selected conversation (navigating back).
     */
    fun closeConversation() {
        _uiState.update {
            it.copy(
                selectedConversation = null,
                messages = emptyList(),
                messagesError = null,
            )
        }
    }

    /**
     * Load and decrypt messages for a conversation.
     */
    fun loadMessages(conversationId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingMessages = true, messagesError = null) }

            try {
                val response = apiService.request<MessagesListResponse>(
                    "GET",
                    "/api/conversations/$conversationId/messages",
                )

                val decrypted = response.messages.mapNotNull { message ->
                    decryptMessage(message)
                }

                _uiState.update {
                    it.copy(
                        messages = decrypted,
                        isLoadingMessages = false,
                        totalMessages = response.total,
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isLoadingMessages = false,
                        messagesError = e.message ?: "Failed to load messages",
                    )
                }
            }
        }
    }

    /**
     * Send an encrypted reply in the current conversation.
     *
     * @param text The plaintext message to send
     */
    fun sendReply(text: String) {
        val conversation = _uiState.value.selectedConversation ?: return
        if (text.isBlank()) return

        viewModelScope.launch {
            _uiState.update { it.copy(isSending = true, sendError = null) }

            try {
                // Collect reader pubkeys: self + assigned volunteer (if different) + admins
                val readerPubkeys = buildList {
                    cryptoService.pubkey?.let { add(it) }
                    conversation.assignedVolunteerPubkey?.let { volunteerPub ->
                        if (volunteerPub != cryptoService.pubkey) add(volunteerPub)
                    }
                    // Include admin pubkeys so admins can decrypt messages
                    sessionState.adminPubkeys.forEach { adminPub ->
                        if (adminPub !in this) add(adminPub)
                    }
                }

                val encrypted = cryptoService.encryptMessage(text, readerPubkeys)

                val envelopes = encrypted.envelopes.map { env ->
                    CreateMessageEnvelope(
                        pubkey = env.recipientPubkey,
                        wrappedKey = env.wrappedKey,
                        ephemeralPubkey = env.ephemeralPubkey,
                    )
                }

                val request = SendMessageRequest(
                    encryptedContent = encrypted.ciphertext,
                    readerEnvelopes = envelopes,
                )

                apiService.request<ConversationMessage>(
                    "POST",
                    "/api/conversations/${conversation.id}/messages",
                    request,
                )

                _uiState.update { it.copy(isSending = false) }

                // Reload messages to include the sent message
                loadMessages(conversation.id)
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSending = false,
                        sendError = e.message ?: "Failed to send message",
                    )
                }
            }
        }
    }

    /**
     * Mark all messages in a conversation as read.
     */
    private fun markAsRead(conversationId: String) {
        viewModelScope.launch {
            try {
                apiService.requestNoContent(
                    "POST",
                    "/api/conversations/$conversationId/read",
                )
            } catch (_: Exception) {
                // Read receipts are non-critical
            }
        }
    }

    /**
     * Clear the send error.
     */
    fun clearSendError() {
        _uiState.update { it.copy(sendError = null) }
    }

    fun dismissError() {
        _uiState.update { it.copy(error = null) }
    }

    // ---- Search ----

    /**
     * Set the search query and filter conversations.
     */
    fun setSearchQuery(query: String) {
        _uiState.update { it.copy(searchQuery = query) }
    }

    /**
     * Get conversations filtered by the current search query.
     */
    fun filteredConversations(): List<Conversation> {
        val query = _uiState.value.searchQuery.lowercase()
        if (query.isBlank()) return _uiState.value.conversations
        return _uiState.value.conversations.filter { conversation ->
            conversation.contactHash.lowercase().contains(query) ||
                    conversation.channelType.lowercase().contains(query)
        }
    }

    // ---- Conversation Actions ----

    /**
     * Close/resolve the currently selected conversation.
     */
    fun closeSelectedConversation() {
        val conversation = _uiState.value.selectedConversation ?: return
        viewModelScope.launch {
            try {
                apiService.requestNoContent(
                    "POST",
                    "/api/conversations/${conversation.id}/close",
                )
                _uiState.update {
                    it.copy(
                        selectedConversation = it.selectedConversation?.copy(status = "closed"),
                    )
                }
                loadConversations()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(sendError = e.message ?: "Failed to close conversation")
                }
            }
        }
    }

    /**
     * Reopen a closed conversation.
     */
    fun reopenSelectedConversation() {
        val conversation = _uiState.value.selectedConversation ?: return
        viewModelScope.launch {
            try {
                apiService.requestNoContent(
                    "POST",
                    "/api/conversations/${conversation.id}/reopen",
                )
                _uiState.update {
                    it.copy(
                        selectedConversation = it.selectedConversation?.copy(status = "active"),
                    )
                }
                loadConversations()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(sendError = e.message ?: "Failed to reopen conversation")
                }
            }
        }
    }

    /**
     * Assign the current conversation to a volunteer.
     */
    fun assignConversation(volunteerPubkey: String) {
        val conversation = _uiState.value.selectedConversation ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(showAssignDialog = false) }
            try {
                apiService.requestNoContent(
                    "POST",
                    "/api/conversations/${conversation.id}/assign",
                    mapOf("volunteerPubkey" to volunteerPubkey),
                )
                _uiState.update {
                    it.copy(
                        selectedConversation = it.selectedConversation?.copy(
                            assignedVolunteerPubkey = volunteerPubkey,
                        ),
                    )
                }
                loadConversations()
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(sendError = e.message ?: "Failed to assign conversation")
                }
            }
        }
    }

    fun showAssignDialog() {
        _uiState.update { it.copy(showAssignDialog = true) }
        loadUsersForAssign()
    }

    fun dismissAssignDialog() {
        _uiState.update { it.copy(showAssignDialog = false) }
    }

    private fun loadUsersForAssign() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingVolunteers = true) }
            try {
                val response = apiService.request<UsersListResponse>(
                    "GET",
                    "/api/users?limit=100",
                )
                _uiState.update {
                    it.copy(
                        assignableVolunteers = response.users.filter { v -> v.status == "active" },
                        isLoadingVolunteers = false,
                    )
                }
            } catch (_: Exception) {
                _uiState.update { it.copy(isLoadingVolunteers = false) }
            }
        }
    }

    /**
     * Decrypt a single message by finding our envelope and calling CryptoService.
     */
    private suspend fun decryptMessage(message: ConversationMessage): DecryptedMessage? {
        val ourPubkey = cryptoService.pubkey ?: return null

        val envelope = message.recipientEnvelopes.find { it.pubkey == ourPubkey }
            ?: return null

        return try {
            val plaintext = cryptoService.decryptMessage(
                encryptedContent = message.encryptedContent,
                wrappedKey = envelope.wrappedKey,
                ephemeralPubkey = envelope.ephemeralPubkey,
            )

            if (plaintext != null) {
                DecryptedMessage(
                    id = message.id,
                    text = plaintext,
                    direction = message.direction,
                    channelType = message.channelType,
                    createdAt = message.createdAt,
                    isRead = message.readAt != null,
                )
            } else {
                null
            }
        } catch (_: Exception) {
            null
        }
    }
}

File: apps/ios/Sources/ViewModels/ConversationsViewModel.swift
import Foundation
import UIKit

// MARK: - ConversationsViewModel

/// View model for the Conversations tab. Fetches conversations from the API,
/// manages status filtering, and handles real-time message updates via WebSocket.
@Observable
final class ConversationsViewModel {
    private let apiService: APIService
    private let cryptoService: CryptoService
    private let webSocketService: WebSocketService
    private let hubContext: HubContext
    private let adminPubkeys: [String]

    // MARK: - Public State

    /// All conversations from the server, filtered by current status filter.
    var filteredConversations: [AppConversation] = []

    /// All conversations (unfiltered), used for badge count calculations.
    var allConversations: [AppConversation] = []

    /// Total unread message count across all conversations.
    var totalUnreadCount: Int {
        allConversations.reduce(0) { $0 + $1.unreadCount }
    }

    /// Current status filter.
    var statusFilter: ConversationStatusFilter = .active {
        didSet { applyFilter() }
    }

    /// Whether the initial load is in progress.
    var isLoading: Bool = false

    /// Error message from the last failed operation.
    var errorMessage: String?

    /// Messages for the currently viewed conversation.
    var currentMessages: [DecryptedMessage] = []

    /// The currently viewed conversation ID.
    var currentConversationId: String?

    /// Whether messages are loading for the current conversation.
    var isLoadingMessages: Bool = false

    /// Text input for the reply field.
    var replyText: String = ""

    /// Whether a send is in progress.
    var isSending: Bool = false

    // MARK: - Private State

    private var eventTask: Task<Void, Never>?

    // MARK: - Initialization

    init(apiService: APIService, cryptoService: CryptoService, webSocketService: WebSocketService, hubContext: HubContext, adminPubkeys: [String] = []) {
        self.apiService = apiService
        self.cryptoService = cryptoService
        self.webSocketService = webSocketService
        self.hubContext = hubContext
        self.adminPubkeys = adminPubkeys
    }

    // MARK: - Data Loading

    /// Load conversations from the API.
    func loadConversations() async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil

        do {
            let response: ConversationsListResponse = try await apiService.request(
                method: "GET",
                path: "/api/conversations"
            )
            allConversations = response.conversations.sorted { lhs, rhs in
                // Sort by last message time, newest first
                let lhsDate = lhs.lastMessageDate ?? lhs.createdDate ?? Date.distantPast
                let rhsDate = rhs.lastMessageDate ?? rhs.createdDate ?? Date.distantPast
                return lhsDate > rhsDate
            }
            applyFilter()
        } catch {
            if case APIError.noBaseURL = error {
                // Hub not configured — show empty state
            } else {
                errorMessage = error.localizedDescription
            }
        }

        isLoading = false
    }

    /// Refresh conversations (pull-to-refresh).
    func refresh() async {
        isLoading = false
        await loadConversations()
    }

    // MARK: - Message Loading

    /// Load messages for a specific conversation and decrypt them.
    func loadMessages(for conversationId: String) async {
        currentConversationId = conversationId
        isLoadingMessages = true
        currentMessages = []

        do {
            let response: ConversationMessagesResponse = try await apiService.request(
                method: "GET",
                path: "/api/conversations/\(conversationId)/messages"
            )

            currentMessages = response.messages.compactMap { decryptConversationMessage($0) }
                .sorted { $0.createdAt < $1.createdAt }

            // Mark as read
            await markAsRead(conversationId: conversationId)
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoadingMessages = false
    }

    // MARK: - Send Reply

    /// Encrypt and send a reply message in the current conversation.
    func sendReply() async {
        guard let conversationId = currentConversationId else { return }
        let text = replyText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }

        isSending = true

        do {
            // Find the conversation to get reader pubkeys
            let conversation = allConversations.first { $0.id == conversationId }
            var readerPubkeys: [String] = []

            // Include our own pubkey for self-decryption
            if let ourPubkey = cryptoService.pubkey {
                readerPubkeys.append(ourPubkey)
            }

            // Include assigned volunteer (may be us or someone else)
            if let assignedPubkey = conversation?.assignedVolunteerPubkey,
               !readerPubkeys.contains(assignedPubkey) {
                readerPubkeys.append(assignedPubkey)
            }

            // Include admin pubkeys so admins can decrypt messages
            for adminPubkey in adminPubkeys where !readerPubkeys.contains(adminPubkey) {
                readerPubkeys.append(adminPubkey)
            }

            let encrypted = try cryptoService.encryptMessage(
                plaintext: text,
                readerPubkeys: readerPubkeys
            )

            let request = SendMessageRequest(
                encryptedContent: encrypted.encryptedContent,
                recipientEnvelopes: encrypted.envelopes
            )

            let _: ConversationMessage = try await apiService.request(
                method: "POST",
                path: "/api/conversations/\(conversationId)/messages",
                body: request
            )

            // Haptic feedback on success
            let generator = UIImpactFeedbackGenerator(style: .light)
            generator.impactOccurred()

            // Clear input and reload messages
            replyText = ""
            await loadMessages(for: conversationId)
        } catch {
            errorMessage = error.localizedDescription
        }

        isSending = false
    }

    // MARK: - Mark as Read

    /// Mark all messages in a conversation as read.
    private func markAsRead(conversationId: String) async {
        do {
            let _: MarkReadResponse = try await apiService.request(
                method: "POST",
                path: "/api/conversations/\(conversationId)/read"
            )

            // Update the local unread count
            if let index = allConversations.firstIndex(where: { $0.id == conversationId }) {
                // Re-fetch to get updated count since Conversation is immutable
                await loadConversations()
            }
        } catch {
            // Non-critical — don't surface this error
        }
    }

    // MARK: - Real-time Events

    /// Start listening for typed WebSocket events related to conversations.
    func startEventListener() {
        eventTask?.cancel()
        eventTask = Task { [weak self] in
            guard let self else { return }
            for await attributed in self.webSocketService.attributedEvents {
                guard !Task.isCancelled else { break }
                guard attributed.hubId == self.hubContext.activeHubId else { continue }
                await self.handleTypedEvent(attributed.event)
            }
        }
    }

    /// Stop listening for events.
    func stopEventListener() {
        eventTask?.cancel()
        eventTask = nil
    }

    /// Handle a decrypted, typed hub event — only react to conversation-related events.
    @MainActor
    private func handleTypedEvent(_ eventType: HubEventType) {
        switch eventType {
        case .messageNew, .conversationAssigned, .conversationClosed:
            Task { await loadConversations() }
            if let id = currentConversationId {
                Task { await loadMessages(for: id) }
            }
        case .conversationNew:
            Task { await loadConversations() }
        case .messageStatus:
            if let id = currentConversationId {
                Task { await loadMessages(for: id) }
            }
        case .callRing, .callAnswered, .callUpdate, .callEnded, .voicemailNew, .presenceSummary, .presenceDetail, .shiftStarted, .shiftEnded, .shiftUpdate, .noteCreated, .unknown:
            // Not conversation-related — ignore
            break
        }
    }

    // MARK: - Message Decryption

    /// Decrypt a conversation message using the recipient envelope for our pubkey.
    private func decryptConversationMessage(_ message: ConversationMessage) -> DecryptedMessage? {
        guard let ourPubkey = cryptoService.pubkey else { return nil }

        // Find our envelope
        guard let ourEnvelope = message.recipientEnvelopes.first(where: { $0.pubkey == ourPubkey }) else {
            return nil
        }

        do {
            let decryptedText = try cryptoService.decryptMessage(
                encryptedContent: message.encryptedContent,
                wrappedKey: ourEnvelope.wrappedKey,
                ephemeralPubkey: ourEnvelope.ephemeralPubkey
            )

            return DecryptedMessage(
                id: message.id,
                text: decryptedText,
                direction: message.direction,
                channelType: message.channelType,
                createdAt: DateFormatting.parseISO(message.createdAt) ?? Date(),
                isRead: message.isRead
            )
        } catch {
            return nil
        }
    }

    // MARK: - Filtering

    /// Apply the current status filter to the conversations list.
    private func applyFilter() {
        switch statusFilter {
        case .all:
            filteredConversations = allConversations
        case .active:
            filteredConversations = allConversations.filter { $0.conversationStatus == .active }
        case .closed:
            filteredConversations = allConversations.filter { $0.conversationStatus == .closed }
        }
    }

    // MARK: - Helpers


    deinit {
        eventTask?.cancel()
    }
}

// MARK: - ConversationStatusFilter

/// Filter options for the conversations list.
enum ConversationStatusFilter: String, CaseIterable, Sendable {
    case active
    case closed
    case all

    var displayName: String {
        switch self {
        case .active: return NSLocalizedString("filter_active", comment: "Active")
        case .closed: return NSLocalizedString("filter_closed", comment: "Closed")
        case .all: return NSLocalizedString("filter_all", comment: "All")
        }
    }
}
